_posts/2017-05-10-polyglot-scientific-computing-with-xtensor.md

```diff
@@ -0,0 +1,239 @@
+
+For numerical code, C-level efficiency can be achieved in different ways. Preferred approaches differ in Python and Julia. Library authors for Python often make use of tools such as cython, or try to write code in a way that is amenable to vectorization with numpy. Julia fully relies on just-in-time compilation and the LLVM stack. These two approaches are not easily reconcilable.
+
+As library authors, we directly work with C++ for the numerical code, and provide bindings fo the target languages. This is a usecase in which the `xtensor` C++ library particularly shines.
+
+`xtensor` is a C++ lazy tensor algebra library providing an API similar to that of numpy. You can check out the [numpy to xtensor cheat sheet](http://xtensor.readthedocs.io/en/latest/numpy.html). More importantly, `xtensor` is an *expression system*, which enables the use of tensor expression backed by arbitrary backends such as
+
+ - in-memory containers
+ - database systems
+ - filesystem operations
+ - data structures from other libraries or programming languages providing a C API.
+
+In other words, `xtensor` enables a uniform way of operating on data regardless of the source. This is also what enabled language bindings such as [xtensor-python](https://github.com/QuantStack/xtensor-python) and [xtensor-julia](https://github.com/QuantStack/xtensor-julia) which wrap native numpy and Julia arrays as xtensor expressions.
+
+![hope](/assets/images/posts/hope.png)
+<div class="img-caption">
+But there is still hope
+</div>
+
+## Learning by the example
+
+For instance, assume we have developped, in Python, some statistics functions based on `scipy.logsumexp`. After a few days in production, we realize that the performances are not so good and that the bottleneck is `scipy.logsumexp`. To fix that, we decide to replace this function with a native implementation, in C++, exposed to Python as an extension module.
+
+## First implementation
+
+Let's start with a straighforward implementation, that is, a simple C++ module exposing its single function to python via `pybind11` and `xtensor-python`:
+
+```cpp
+#define FORCE_IMPORT_ARRAY
+#include "xtensor-python/pyarray.hpp"
+#include "pybind11/stl.h"
+#include "xtensor/xeval.hpp"
+#include "xtensor/xstridedview.hpp"
+
+namespace py = pybind11;
+using array_type = xt::pyarray<double>;
+using axes_type = std::vector<double>;
+
+// C++ implementation of logsumexp using xtensor
+array_type logsumexp(const array_type& a, const axes_type& axes)
+{
+    auto m = xt::eval(xt::amax(a, axes));
+    auto sv = xt::slice_vector(m);
+    for (auto i = 0; i < e1.dimension(); ++i)
+    {
+        if (std::find(axes.begin(), axes.end(), i) != axes.end())
+        {
+            sv.push_back(xt::newaxis());
+        }
+        else
+        {
+            sv.push_back(xt::all());
+        }
+    }
+    auto max2 = xt::dynamic_view(m, sv);
+    return array_type(m + xt::log(xt::sum(xt::exp(a - max2), axes)));
```

Comment:
I quick look at this code, and I wonder 'why all this code'? Isn't there a cleaner way, is the first part needed, can we get rid of that?